File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
          return 0 
        if len(nums) == 1:
          return nums[0]

        ## dp[i] = max money robber can get  if we consider till ith house 

        ## dp[i] = max(dp[i-1], dp[i-2] + nums[i])

        prev2 = nums[0] # for tracking i-2th max money, consider first house

        prev1 = max(nums[0], nums[1]) # consider either first or 2nd house.

        for i in range(2, len(nums)):
          current = max(prev1, prev2 + nums[i])
          prev2 = prev1
          prev1 = current
        return prev1

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.empty()){
          return 0;
        }
        int prev2 = 0;
        int prev1 = 0;
        for (int i = 0; i < nums.size(); i++)
        {
          int take = prev2 + nums[i];
          int skip = prev1;
          int current = max(take, skip);

          prev2 = prev1;
          prev1 = current;


        }
        int result = prev1;

        return result;
    }
};